faiss/IndexNSG.cpp

```diff
@@ -0,0 +1,170 @@
+          storage(storage),
+          is_built(false) {}
+
+IndexNSG::~IndexNSG() {
+    if (own_fields) {
+        delete storage;
+    }
+}
+
+void IndexNSG::train(idx_t n, const float* x) {
+    FAISS_THROW_IF_NOT_MSG(
+            storage,
+            "Please use IndexNSGFlat (or variants) instead of IndexNSG directly");
+    // nsg structure does not require training
+    storage->train(n, x);
+    is_trained = true;
+}
+
+void IndexNSG::search(
+        idx_t n,
+        const float* x,
+        idx_t k,
+        float* distances,
+        idx_t* labels) const
+
+{
+    FAISS_THROW_IF_NOT_MSG(
+            storage,
+            "Please use IndexNSGFlat (or variants) instead of IndexNSG directly");
+
+    idx_t check_period = InterruptCallback::get_period_hint(d * nsg.search_L);
+
+    for (idx_t i0 = 0; i0 < n; i0 += check_period) {
+        idx_t i1 = std::min(i0 + check_period, n);
+
+#pragma omp parallel
+        {
+            VisitedTable vt(ntotal);
+
+            DistanceComputer* dis = storage_distance_computer(storage);
+            ScopeDeleter1<DistanceComputer> del(dis);
+
+#pragma omp for
+            for (idx_t i = i0; i < i1; i++) {
+                idx_t* idxi = labels + i * k;
+                float* simi = distances + i * k;
+                dis->set_query(x + i * d);
+
+                maxheap_heapify(k, simi, idxi);
+                nsg.search(*dis, k, idxi, simi, vt);
+                maxheap_reorder(k, simi, idxi);
+
+                vt.advance();
+            }
+        }
+        InterruptCallback::check();
+    }
+
+    if (metric_type == METRIC_INNER_PRODUCT) {
+        // we need to revert the negated distances
+        for (size_t i = 0; i < k * n; i++) {
+            distances[i] = -distances[i];
+        }
+    }
+}
+
+void IndexNSG::build(idx_t n, const float* x, idx_t* knn_graph, int GK) {
+    FAISS_THROW_IF_NOT_MSG(
+            storage,
+            "Please use IndexNSGFlat (or variants) instead of IndexNSG directly");
+    FAISS_THROW_IF_NOT_MSG(!is_built, "The IndexNSG is already built");
+    storage->add(n, x);
+    ntotal = storage->ntotal;
+
+    nsg::Graph<idx_t> knng(knn_graph, n, GK);
+
+    // check the knn graph
+    idx_t count = 0;
+    for (idx_t i = 0; i < n; i++) {
+        for (int j = 0; j < GK; j++) {
+            idx_t id = knng.at(i, j);
+            if (id < 0 || id >= n) {
+                count += 1;
+            }
+        }
+    }
+    if (count > 0) {
+        fprintf(stderr,
+                "WARNING: the input knn graph "
+                "has %ld invalid entries\n",
+                count);
+    }
+
+    nsg.build(storage, n, knng, verbose);
+    is_built = true;
+}
+
+void IndexNSG::add(idx_t n, const float* x) {
+    // FAISS_THROW_IF_NOT_MSG(storage,
+    //    "Please use IndexNSGFlat (or variants) instead of IndexNSG directly");
+    // FAISS_THROW_IF_NOT(is_trained && is_built);
+    // int n0 = ntotal;
+    // storage->add(n, x);
+    // ntotal = storage->ntotal;
+
```

Comment:
Please implement a default add function with brute-force search. 
2nd step: support nndescent via a flag that switches between the two. 
Callers can still call `build()` themselves if they want to do it otherwise (eg. with GPU).